File: src/main/java/xyz/devcmb/treeTumblers/util/Format.java
package xyz.devcmb.treeTumblers.util;


import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextColor;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.NotNull;
import xyz.devcmb.treeTumblers.data.DataManager;

import java.util.Map;

public class Format {
    public static Component format(String string, FormatType type) {
        Map<String, Object> data = type.getData();
        return Component.text(
                data.get("prefix").toString()
        ).append(Component.text(" ")
        .append(Component.text(string).color((NamedTextColor)data.get("color"))));
    }

    public static Component format(Component component, FormatType type) {
        Map<String, Object> data = type.getData();
        return Component.text(
                data.get("prefix").toString()
        ).append(Component.text(" ")
                .append(component));
    }

    public enum FormatType {
        SUCCESS(Map.of(
            "prefix", "✅",
            "color", NamedTextColor.GREEN
        )),
        WARNING(Map.of(
                "prefix", "⚠️",
                "color", NamedTextColor.YELLOW
        )),
        ERROR(Map.of(
                "prefix", "❌",
                "color", NamedTextColor.DARK_RED
        )),
        INVALID(Map.of(
                "prefix", "❓",
                "color", NamedTextColor.RED
        )),
        INFO(Map.of(
                "prefix", "ℹ️",
                "color", NamedTextColor.AQUA
        ));

        private final Map<String, Object> data;
        FormatType(Map<String, Object> map) {
            data = map;
        }

        public Map<String, Object> getData() {
            return data;
        }
    }

    public static String stringToUnicode(String string){
        return Character.toString((char) Integer.parseInt(string.replace("\\u", ""), 16));
    }

    public static Component formatPlayerName(@NotNull OfflinePlayer player){
        if(player.getName() == null) return Component.empty();
        Component formatted;

        DataManager.TeamData teamData = DataManager.GetPlayerTeam(player);
        if(teamData == null){
            formatted = Component.empty().append(
                    Component.text(player.getName()).color(NamedTextColor.GRAY)
            );

            return formatted;
        }

        String iconChar = Character.toString((char) Integer.parseInt(teamData.icon.replace("\\u", ""), 16));
        formatted = Component.empty().append(Component.empty().append(Component.text(iconChar + " ").color(NamedTextColor.WHITE)).font(Key.key("tumbling:icons"))).append(Component.text(player.getName()).color(TextColor.fromHexString(teamData.color)));

        return formatted;
    }
}

File: src/main/kotlin/xyz/devcmb/treeTumblers/util/Format.kt
package xyz.devcmb.treeTumblers.util

import net.kyori.adventure.key.Key
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextColor
import org.bukkit.OfflinePlayer
import xyz.devcmb.treeTumblers.TreeTumblers
import xyz.devcmb.treeTumblers.data.DataManager
import java.util.Map

object Format {
    fun format(string: String, type: FormatType): Component {
        val data = type.data
        return Component.text(
            data["prefix"].toString()
        ).append(
            Component.text(" ")
                .append(Component.text(string).color(data["color"] as NamedTextColor?))
        )
    }

    fun format(component: Component, type: FormatType): Component {
        val data = type.data
        return Component.text(
            data["prefix"].toString()
        ).append(
            Component.text(" ")
                .append(component)
        )
    }

    fun stringToUnicode(string: String): String {
        return string.replace("\\u", "").toInt(16).toChar().toString()
    }

    fun formatPlayerName(player: OfflinePlayer): Component {

        return Component.empty()
    }

    fun formatTeamName(team: String?): Component {
        if(team == null) {
            // TODO: add spectator icon
            return Component.empty()
                .append(Component.text("Spectator").color(NamedTextColor.DARK_GRAY))
        }

        val data = DataManager.teamData[team]
        if(data == null) {
            TreeTumblers.LOGGER.warning("⚠️ | Can not format a team name for a nonexistent team ($team)")
            return Component.empty()
        }

        return Component.empty()
            .append(Component.text(stringToUnicode(data.icon)).font(Key.key("tumbling:icons")).color(NamedTextColor.WHITE))
            .append(Component.text(" ${data.displayName}").color(TextColor.fromHexString(data.color)))
    }

    enum class FormatType(val data: MutableMap<String?, Any?>) {
        SUCCESS(
            Map.of<String?, Any?>(
                "prefix", "✅",
                "color", NamedTextColor.GREEN
            )
        ),
        WARNING(
            Map.of<String?, Any?>(
                "prefix", "⚠️",
                "color", NamedTextColor.YELLOW
            )
        ),
        ERROR(
            Map.of<String?, Any?>(
                "prefix", "❌",
                "color", NamedTextColor.DARK_RED
            )
        ),
        INVALID(
            Map.of<String?, Any?>(
                "prefix", "❓",
                "color", NamedTextColor.RED
            )
        ),
        INFO(
            Map.of<String?, Any?>(
                "prefix", "ℹ️",
                "color", NamedTextColor.AQUA
            )
        )
    }
}